src/backend/common/Builder.h

```diff
@@ -0,0 +1,100 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BACKEND_COMMON_BUILDER_H_
+#define BACKEND_COMMON_BUILDER_H_
+
+#include "Forward.h"
+#include "RefCounted.h"
+
+#include "nxt/nxtcpp.h"
+
+#include <string>
+
+namespace backend {
+
+    // This class implements behavior shared by all builders:
+    //  - Tracking whether GetResult has been called already, needed by the
+    //    autogenerated code to prevent operations on "consumed" builders.
+    //  - The error status callback of the API. The callback is guaranteed to be
+    //    called exactly once with an error, a success, or "unknown" if the
+    //    builder is destroyed; also the builder callback cannot be called before
+    //    either the object is destroyed or GetResult is called.
+    //
+    // It is possible for error to be generated before the error callback is
+    // registered when a builder "set" function performance validation inline.
+    // Because of this we have to store the status in the builder and defer
+    // calling the callback to GetResult.
+
+    class Builder : public RefCounted {
```

Comment:
Seems like you're gonna have the same `GetResult` boilerplate everywhere; could defined in `Builder` instead if it were templated? `GetResultImpl` would be a virtual on `Builder`, `HandleResult` would use the template of the class.